api/test/unit/views/test_image_views.py

```diff
@@ -79,3 +80,31 @@ def test_thumbnail_uses_upstream_thumb_for_smk(
         thumb_call.return_value = mock_response
         api_client.get(f"/v1/images/{image.identifier}/thumb/")
     thumb_call.assert_called_once_with(ANY, image, expected_thumb_url)
+
+
+@pytest.mark.django_db
+def test_watermark_raises_424_for_invalid_image(api_client):
+    image = ImageFactory.create()
+    expected_error_message = (
+        "cannot identify image file <_io.BytesIO object at 0xffff86d8fec0>"
+    )
+
+    with patch("PIL.Image.open") as mock_open:
+        mock_open.side_effect = UnidentifiedImageError(expected_error_message)
+        res = api_client.get(f"/v1/images/{image.identifier}/watermark/")
+    assert res.status_code == 424
+    assert res.data["detail"] == expected_error_message
```

Comment:
This isn't quite testing that the exception is raised. I'd suggest something using `pytest.raises` like the following though it might need more work. Could you look into it? Same for the last test.
 
```suggestion
@pytest.mark.django_db
def test_watermark_raises_424_for_invalid_image(api_client):
    image = ImageFactory.create()
    with patch(
        "PIL.Image.open", side_effect=UnidentifiedImageError("bad image")
    ), pytest.raises(UpstreamWatermarkException):
        res = api_client.get(f"/v1/images/{image.identifier}/watermark/")
    assert res.status_code == 424
```

Reply:
@krysal, I tried using the `pytest.raises` as you suggested, but this approach fails. So, I tried to find a reason for this. It seems that DRF catches `ApiException`s and converts them to a response with the relevant error/status code:

https://github.com/encode/django-rest-framework/blob/f56b85b7dd7e4f786e0769bba6b7609d4507da83/rest_framework/views.py#L448

So, the previous test that checks for the `status_code` and the `message` is actually the only way to see which exception was raised.